### ISSUE TITLE
Cast matrices as vectors?

### ISSUE THREAD

**Issue by cdeterman** [viennacl/viennacl-dev#207]
Is there any way to convert an existing `viennacl::matrix` to a `viennacl::vector`?  There appears to not be an operator for doing it directly with:

``` c++
viennacl::matrix<SCALAR_TYPE> mat;
viennacl::vector<SCALAR_TYPE> vec = mat;
```

Is there a more roundabout way perhaps such as initializing the vector and filling it?  For example, I can think of grabbing columns/rows one at a time but not quite sure how to put in incremental position within the initialized vector.

**Comment by karlrupp** [comment]
Not for a `vector<T>`, but you can use `vector_base<T>`:

    viennacl::matrix<T> A(...);
    viennacl::vector_base<T> x(A.handle(), A.internal_size(), 0, 1);

However, keep the padding of A in mind. Also, `x` will be untangled from `A` if you resize any of the two.

**Comment by cdeterman** [comment]
Thanks, by 'untangled' do you mean that the data would be copied?  If the no resizing takes place I assume then that the data is shared?  I believe this is what you mean and the way the code looks but I want to be clear.

**Comment by karlrupp** [comment]
Yes, data is shared until you resize. If you resize either of the objects, a new memory buffer will be allocated and used instead of the 'old' shared buffer. However, this new buffer is only used by the object on which you resized, so the other object will keep its old data and sizes.

The reason for this behavior is to have consistent object states at all time. For example, if you resize the vector having shared data with a 3x4-matrix to length 13, it is not clear what a new buffer of size 13 means for the matrix.

**Comment by cdeterman** [comment]
Thanks Karl, that satisfies this question.

**Comment by cdeterman** [comment]
It occurred to me, is the opposite possible as well?  Can I cast a vector as a matrix?  It doesn't necessarily need to share the memory but that, of course, would be ideal.

**Comment by karlrupp** [comment]
Yes, you can do that in a similar way: Pass the memory handle `x.handle()` to the respective constructor of `matrix_base`, where you also specify the row and column dimensions (including 'internal' dimensions). :-)

**Comment by cdeterman** [comment]
Will that fill the expected matrix elements though?  If sharing the memory I would guess that it may extend elements in to the 'padded' elements.

**Comment by karlrupp** [comment]
you can 'disable' the padding by setting the number of rows equal to the number of internal rows (and similar for columns). This will create a matrix with 'shared' data. If you want to copy instead of share, pass that shared matrix to the copy-constructor of a new matrix.

**Comment by cdeterman** [comment]
The padding is fine, if I understand correctly it exists for performance reasons, I just want to make sure that the values go in the correct spots.  For example a vector containing 1,2,3,4,5,6,7,8 put in to a 2x4 matrix doesn't end up like this:

1 2 3 4 5 6 7 8 0 0 0 ...
0 0 0 0 ...

but instead still end up like this:

1 2 3 4 0 0 0 0 ...
5 6 7 8 0 0 0 0 ...

**Comment by cdeterman** [comment]
For example, the following results in an error:

```c++
include <iostream>

#define VIENNACL_WITH_OPENCL 1

#include "viennacl/matrix.hpp"
#include "viennacl/vector.hpp"


int main(){

  viennacl::vector<float> v = viennacl::scalar_vector<float>(8, 2); 

  viennacl::matrix<float> r = viennacl::zero_matrix<float>(2,4);

  viennacl::matrix_base<float> m(v.handle(), 2, 0, 0, r.internal_size1(), 4, 0, 0, r.internal_size2(), true);

  std::cout << v << std::endl;

  std::cout << r << std::endl;

  std::cout << m << std::endl;  

}
```

```
[8](2,2,2,2,2,2,2,2)
[2,4]((0,0,0,0),(0,0,0,0))
terminate called after throwing an instance of 'viennacl::ocl::invalid_value'
  what():  ViennaCL: FATAL ERROR: CL_INVALID_VALUE.
If you think that this is a bug in ViennaCL, please report it at viennacl-support@lists.sourceforge.net and supply at least the following information:
 * Operating System
 * Which OpenCL implementation (AMD, NVIDIA, etc.)
 * ViennaCL version
Many thanks in advance!
Aborted (core dumped)
```

What am I doing wrong here?  I'm not sure about the copy-constructor you are mentioning either.

**Comment by karlrupp** [comment]
The constructor call for `m` should be:

    m(v.handle(),
        2, 0, 1, 2,    // 2 rows, start 0, inc 1, internal size 2
        4, 0, 1, 4,   //  4 cols, start 0, inc 1, internal size 4
        true);

http://viennacl.sourceforge.net/doc/manual-types.html#manual-types-matrix

**Comment by cdeterman** [comment]
@karlrupp that does indeed get the casting to work as you describe.  Should I care about setting up the matrix with no padding?  I thought it was preferred to have the matrices padded.

Second, is the link you provided to address the copy-constructor?  I see the familiar `copy` function but I don't see anything about a 'copy-constructor' in the event I don't what the memory shared.

Thanks

**Comment by cdeterman** [comment]
@karlrupp The original question I posed is also potentially a problem.  When casting the matrix as a vector, the padded elements are put in the resulting vector.  I understand this from a memory sharing standpoint but what about the situation where I don't want shared memory (as noted in my previous comment).  How can I cast the elements within size1 and size2 to a vector.

**Comment by cdeterman** [comment]
I also see the vector constructor from a pointer, can that be used to keep the same viennacl::vector type?

http://viennacl.sourceforge.net/doc/classviennacl_1_1vector.html

vector (NumericT *ptr_to_mem, viennacl::memory_types mem_type, size_type vec_size, size_type start=0, size_type stride=1)

Or should I be considering changing a lot of my code to use the *_base objects?

**Comment by karlrupp** [comment]
Well, you can create an unpadded matrix, copy from the padded matrix, and then instantiate a vector sharing the data with the padded matrix.

What do you mean by 'keep the same viennacl::vector type'? If you want to avoid using vector_base: No, there's no way around vector_base if you want to support all three compute backends.

**Comment by cdeterman** [comment]
@karlrupp there are two last things I am not clear on here.

1. How to create an unpadded matrix.  Is there any documentation on how to accomplish this?
2. How to copy only the 'active' elements from a padded matrix to a vector object so that it is only the length of `mat.size1() * mat.size2()`.  **Note - this doesn't need to share memory.**

**Comment by karlrupp** [comment]
ad 1.) this is currently not possible. It is an oversight and will be fixed asap. Thanks for bringing this to my attention
ad 2.) The following should work:

    viennacl::vector<T> x(mat.size1() * mat.size2());
    viennacl::matrix_base<T> dummy(x.handle(),
                                   mat.size1(), 0, 1, mat.size1(),   //row layout
                                   mat.size2(), 0, 1, mat.size2(),   //column layout
                                   true); // row-major
    dummy = mat;

At the end of the code, x will hold the entries of mat without padding, row-by-row. `dummy` is only a wrapper around x to specify the row- and column- layout.

**Comment by cdeterman** [comment]
@karlrupp You have helped me dramatically with your responses but I have yet one more related problem here that I didn't feel required a new issue.  I want to assign a vector to the 'active' matrix elements.  Previously I have done this:

```c++
    viennacl::matrix<T> M;
    viennacl::vector_base<T> V;
    
    viennacl::matrix_base<T> dummy(V.handle(),
                                   M.size1(), 0, 1, M.size1(),   //row layout
                                   M.size2(), 0, 1, M.size2(),   //column layout
                                   true); // row-major
    M = dummy;
```

This works perfectly, however, this fails for when using `vector_range` objects where the range is anywhere but at the beginning because the handle (I believe) still points to the beginning of the vector

```c++
    viennacl::matrix<T> M;
    viennacl::vector_range<viennacl::vector_base<T> > V;
    
    viennacl::matrix_base<T> dummy(V.handle(),
                                   M.size1(), 0, 1, M.size1(),   //row layout
                                   M.size2(), 0, 1, M.size2(),   //column layout
                                   true); // row-major
    M = dummy;
```

Any thoughts on how this could also be accomplished?

**Comment by karlrupp** [comment]
Yes, for a vector range this is not going to work. I can't think of a general way of accomplishing what you want. I recommend copying the vector_range to a new (temporary) vector and then use a `dummy` matrix object for that temporary vector as before. This will have a bit of overhead, though.

I'll let you know once I have a full API ready.

**Comment by cdeterman** [comment]
@karlrupp You say you will notify when you have an API ready for these different casts so ranges can be updated.  Do you have plans on how you can accomplish this functionality that you say that or is it still in planning stages?  I am actually dealing with the reverse problem where I am trying to update a `vector_range` with a `viennacl::matrix` object.  The vectors and matrices do become larger and are a significant overhead (at least with my current card).  

Essentially what I am asking, in the most respectful terms, is if this is something I will need to live with for an extended period of time or should I expect to see enhancements in the near future?  I know you are likely quite busy but this issue is of particular interest to me now at this point.

**Comment by karlrupp** [comment]
@cdeterman  My intention is to look at how MATLAB deals with the problem and then code up a similar API. This will most likely result in an intuitive and easy-to-use API for the user.

Currently I'm very busy with non-ViennaCL work, so it won't happen in 2016. The long overdue 1.8.0 release will be in Q1 2017; I should be able to squeeze it in there. I reoben this pull request as a reminder.
